Add Alpaka implementation of PFRecHitProducer

fllor opened the pull request:
#### PR description:

This pull-request is the first step towards making the ParticleFlow algorithm portable and heterogeneous using Alpaka. It adds a PFRecHitSoA data format and a producer that constructs it. The module can process HCAL and ECAL hits and runs on the CPU and CUDA backends of Alpaka. The ROCm backend compiles, but is untested.
Since calorimeter reconstruction is not available using Alpaka yet, this PR also introduces a CaloRecHitSoA format and a producer that converts calorimeter hits from the CPU modules to the SoA. This format and producer will become obsolete, when then HCAL/ECAL groups directly provide their own SoA formats.
A producer that converts to new PFRecHitSoA format back to the existing PFRecHit collection format is included. It can be used for validation and to run down-stream modules.
This PR does not alter any existing workflow in any way. The Alpaka module is provided but not executed for HLT reconstruction. For now, it can be used for testing and for down-stream development. The transition from the CPU implementation to Alpaka is intended to happen later, when an Alpaka implementation of the PFClusterProducer is also available.

#### PR validation:

Extensive validation of the Alpaka implementation is provided, see commit message of 3f4d1235b92a1151179071f3e1396ceb34d50304.

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

This is not a backport.

@hatakeyamak @jsamudio

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42807/36921

- This PR adds an extra 204KB to repository

cmsbuild commented:
A new Pull Request was created by  @fllor  for master.

It involves the following packages:

- DataFormats/ParticleFlowReco (**reconstruction**)
- RecoParticleFlow/PFRecHitProducer (****)


The following packages do not have a category, yet:

RecoParticleFlow/PFRecHitProducer
Please create a PR for https://github.com/cms-sw/cms-bot/blob/master/categories_map.py to assign category

@cmsbuild, @jfernan2, @mandrenguyen can you please review it and eventually sign? Thanks.
@mmarionncern, @rchatter, @wang0jin, @rovere, @argiro, @missirol, @thomreis, @hatakeyamak, @lgray, @seemasharmafnal this is something you requested to watch as well.
@rappoccio, @antoniovilela, @sextonkennedy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

fllor commented:
> The following packages do not have a category, yet:
> RecoParticleFlow/PFRecHitProducer Please create a PR for https://github.com/cms-sw/cms-bot/blob/master/categories_map.py to assign category

PR created: cms-sw/cms-bot/pull/2070

makortel commented:
assign heterogeneous

cmsbuild commented:
New categories assigned: heterogeneous

@fwyzard,@makortel you have been requested to review this Pull request/Issue and eventually sign? Thanks

hatakeyamak commented:
It may be worthwhile to mention that @filor will talk about the work that went into this PR tomorrow in the particle flow group meeting.
https://indico.cern.ch/event/1326067/
@swagata87 [I don't know Riccardo's github username.]

makortel commented on RecoParticleFlow/PFRecHitProducer/plugins/alpaka/PFRecHitProducerKernel.dev.cc:
```diff
@@ -0,0 +1,199 @@
+    alpaka::memset(queue, denseId2pfRecHit_, 0xff);  // Reset denseId -> pfRecHit index map
+    alpaka::memset(queue, num_pfRecHits_, 0x00);     // Reset global pfRecHit counter
+
+    // Use only one block on the synchronous CPU backend, because there is no
+    // performance gain in using multiple blocks, but there is a significant
+    // penalty due to the more complex synchronisation.
```
Could you elaborate this comment? CPU serial backend should just loop serially over the blocks, and thus not have any special synchronization. Or does the comment anticipate a CPU parallel backend (e.g. TBB)?

fllor replied:
The first kernel requires global synchronisation, which is slow when using multiple blocks. When using the serial CPU backend, there is no performance advantage in using multiple blocks, but a significant penalty due to the synchronisation. Therefore, I use a single block for the CPU backend and corresponding synchronisation (this also addresses you `::Blocks` vs `::Grids` comment above). But you are right, the CPU backend does not require any synchronisation, so I can skip the atomic operation in this case.

makortel replied:
I'm actually still confused of my `::Blocks` vs `::Grids` question :) IIUC you want the `j` on line 27 to have unique values for every thread within a grid. IIUC this would correspond the `::Blocks` setting. AFAICT for CPU Serial the `::Blocks` synchronization level on atomics has zero cost (whereas `::Grids` has a cost). Or do you really need to synchronize across grids?

fllor replied:
Yes, `j` is the index of the newly constructed PFRecHit. So it has to be unique and consecutive across all threads. That is why the kernel requires global synchronisation, which is achieved with the `::Grids` parameter in the atomic operation. I found that the `::Grids` synchronisation has a significant cost in the serial CPU backend (as you say).
My solution was to use a single block for the CPU backend and switch to `::Blocks` synchronisation. The synchronisation is not necessary, logically it should be there (and you say it has zero cost).
Alternatively, one could skip the atomic operation for the CPU backend altogether (since it is serial anyway) and use multiple blocks also for the CPU. This would however not work with a potential parallel CPU backend.

I suggest to do the following: Keep the layouts and synchronisation as they are, but replace `cms::alpakatools::requires_single_thread_per_block_v<Acc1D>` by `std::is_same_v<Device, alpaka::DevCpu>` in lines 27 and 162. For the serial CPU backend, it makes no difference anyway, but this should also work reasonably with a potential parallel CPU backend.

makortel replied:
> Yes, `j` is the index of the newly constructed PFRecHit. So it has to be unique and consecutive across all threads. That is why the kernel requires global synchronisation, which is achieved with the `::Grids` parameter in the atomic operation

Shouldn't `::Blocks` be sufficient to provide synchronization across threads of a block?

fllor replied:
Yes, but the GPU backend uses multiple blocks. Synchronisation within each block is not sufficient. That is why we need the more complex `::Grids` synchronisation there.
For the CPU backend, we use a single block, so `::Blocks` is appropriate.

makortel replied:
I fully agree this piece needs synchronization across blocks. But my impression of the Alpaka atomic operation synchronization hierarchy (from quick poking of alpaka code few weeks ago) is
* `::Theads` provides synchronization across threads of a block
* `::Blocks` provides synchronization across (threads of) blocks of a grid (i.e. kernel)
* `::Grids` provides synchronization across (threads of blocks of) grids (i.e. kernels)

If this is the case, then I'd think `::Blocks` would be sufficient for your case. Do you have any pointer to documentation or experience that the behavior of `::Blocks` is something different?

fllor replied:
I am not sure where I got the `::Blocks` and `::Grids` from and I also cannot find any documentation right now. If what you say is correct, `::Blocks` should be changed to `::Threads` for the CPU and `::Grids` to `::Blocks` for the GPU.
I tried changing `::Grids` to `::Blocks` and ran lots of validation of the GPU backend. It validated fine, so you are probably correct. (Then again, I also tried using `::Threads` (which should not work) and it still validated ok, so I am not completely sure)

makortel replied:
(assuming my understanding of the atomic operation hierarchy levels is correct) I'd suggest to just use `::Blocks` for all backends, i.e. remove the special treatment of the CPU backend.

hatakeyamak replied:
@fllor What do you think about @makortel 's suggestions?

makortel commented on RecoParticleFlow/PFRecHitProducer/plugins/alpaka/PFRecHitTopologyESProducer.cc:
```diff
@@ -0,0 +1,215 @@
+      }
+
+      // Remove neighbours that are not backward compatible (only for HCAL)
+      if (std::is_same_v<CAL, HCAL>)
```
```suggestion
      if constexpr (std::is_same_v<CAL, HCAL>) {
```
?
(I'd suggest to add the curly braces to explicitly denote the code blocks for clarity)

makortel commented on RecoParticleFlow/PFRecHitProducer/test/harvestValidationPlot.py:
```diff
@@ -0,0 +1,28 @@
+process.load('FWCore.MessageService.MessageLogger_cfi')
+process.load('Configuration.StandardSequences.DQMSaverAtRunEnd_cff')
+
+process.maxEvents = cms.untracked.PSet(
+    input = cms.untracked.int32(1),
+    output = cms.optional.untracked.allowed(cms.int32,cms.PSet)
+)
```
This could be just
```suggestion
process.maxEvents.input = 1
```

swagata87 commented:
FYI @cms-sw/pf-l2  @bellan

fwyzard commented on DataFormats/ParticleFlowReco/BuildFile.xml:
```diff
@@ -17,8 +17,11 @@
 <use name="rootcore"/>
 <use name="rootmath"/>
-<flags LCG_DICT_HEADER="classes_1.h classes_2.h"/>
-<flags LCG_DICT_XML="classes_def_1.xml classes_def_2.xml"/>
```
A fix https://github.com/cms-sw/cmsdist/pull/8705 should be available in CMSSW_13_3_X_2023-09-19-2300 .

Can you try to undo the BuildFile change ?

fllor replied:
I switched to the release you mention and undid the changes to the build file and class definitions. Now it compiles, runs and validates fine.

fwyzard replied:
@fllor I still see these changes in the PR.
could you push the updated version, where you reverted the BuildFile changes ?

fllor commented on RecoParticleFlow/PFRecHitProducer/interface/PFRecHitParamsHostCollection.h:
```diff
@@ -0,0 +1,12 @@
+
+namespace reco {
+  using PFRecHitHCALParamsHostCollection = PortableHostCollection<PFRecHitHCALParamsSoA>;
+  using PFRecHitECALParamsHostCollection = PortableHostCollection<PFRecHitECALParamsSoA>;
```
Yes, they are implementation details that are only intended for this package. I will add a comment.

fwyzard commented on DataFormats/ParticleFlowReco/BuildFile.xml:
```diff
@@ -17,8 +17,11 @@
+<use name="DataFormats/Portable"/>
+<use name="DataFormats/SoATemplate"/>
+<use name="HeterogeneousCore/AlpakaInterface"/>
+<flags ALPAKA_BACKENDS="1"/>
```
I suspect that with the latest nightly, this could also be
```suggestion
<flags ALPAKA_BACKENDS="!serial"/>
```
but I have not tried it yet.

So I would go with Matti's suggestion.

fwyzard replied:
In the meantime `!serial` is being used by other PRs, so I think
```suggestion
<flags ALPAKA_BACKENDS="!serial"/>
```
is the best way forward.

fllor commented on RecoParticleFlow/PFRecHitProducer/plugins/PFRecHitProducerTest.cc:
```diff
@@ -0,0 +1,312 @@
+#include <utility>
+#include <variant>
+
+class PFRecHitProducerTest : public DQMEDAnalyzer {
```
It is only for testing. I moved it to the test directory, but now the configuration cannot find it anymore. What else do I have to do, to make this work?

fllor replied:
I was missing the BuildFile. I added it and now it works as before.

swagata87 commented:
type pf

fllor commented on RecoParticleFlow/PFRecHitProducer/plugins/alpaka/PFRecHitProducerKernel.dev.cc:
```diff
@@ -0,0 +1,199 @@
+
+  template <typename CAL>
+  PFRecHitProducerKernel<CAL>::PFRecHitProducerKernel(Queue& queue)
+      : denseId2pfRecHit_(cms::alpakatools::make_device_buffer<uint32_t[]>(queue, CAL::SIZE)),
+        num_pfRecHits_(cms::alpakatools::make_device_buffer<uint32_t>(queue)),
```
Yes, they hold temporary event-specific data. I was trying to avoid frequent memory allocations, by keeping these buffers between events. During my testing, there was no notable performance difference though.
When I go with your suggestion, the `kernel` object in `PFRecHitProducerAlpaka::produce` can be made a local variable. This also means that the module can be made a global producer. If the kernel object is created separately for each event anyway, the buffer allocations might as well be kept in the constructor.

fllor replied:
With the new benchmarking method (running the producer multiple times per event), I can measure the performance impact of this issue. Going from the current version (memory buffer lives forever) to your suggestion (allocate buffer for each event) increases the runtime on the GPU backend from ~9us/event to ~16us/event (for HCAL). This is negligible compared to the total HLT time, so I think it is better to stick with your suggestion and have cleaner code.

hatakeyamak commented:
@fllor Do you think you addressed all suggestions from @makortel et al? If so, it will be good to explicitly clarify that.

hatakeyamak commented:
BTW, we should also adopt GT based HCAL threshold in the alpaka version too. Maybe in a follow-up PR.
@fllor @jsamudio

fwyzard commented:
@fllor I will send you some commits to address some technical and stylish issues:
  - fix the dictionaries under `DataFormats/ParticleFlowReco`;
  - when the host types are in the `reco` namespace, move the device types to the `ALPAKA_ACCELERATOR_NAMESPACE::reco` namespace;
  - make the host types available in the `ALPAKA_ACCELERATOR_NAMESPACE::reco` namespace; 
  - clean up the header guards;
  - update the names of structs and constants to follow the CMS coding rules (types are in Uppercase, variables and constants are in lowercase).

fwyzard commented:
Please pick the three commits from https://github.com/fwyzard/cmssw/tree/pfRecHitProducerAlpaka .

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42807/37226

- This PR adds an extra 11528KB to repository

- Found files with invalid states:
  - DataFormats/ParticleFlowReco/src/classes_def_2.xml:
    - Deleted: 42fd4a73bb03b04983328ba453595f9bcc3045b6
    - Added: 3da2bd307442e9edcda5b69cc4e343954465f814
  - DataFormats/ParticleFlowReco/src/classes_1.h:
    - Deleted: 42fd4a73bb03b04983328ba453595f9bcc3045b6
    - Added: 3da2bd307442e9edcda5b69cc4e343954465f814
  - DataFormats/ParticleFlowReco/src/classes_2.h:
    - Deleted: 42fd4a73bb03b04983328ba453595f9bcc3045b6
    - Added: 3da2bd307442e9edcda5b69cc4e343954465f814
  - DataFormats/ParticleFlowReco/src/alpaka/classes_serial.h:
    - Added: 71059c160a31476481edef20b1f3763d0a88045d
    - Modified: 3899652581347d9d54e2a3ca036c6dfba514c5c4
    - Deleted: 3da2bd307442e9edcda5b69cc4e343954465f814
  - DataFormats/ParticleFlowReco/src/classes_def.xml:
    - Added: 42fd4a73bb03b04983328ba453595f9bcc3045b6
    - Modified: 71059c160a31476481edef20b1f3763d0a88045d, 3899652581347d9d54e2a3ca036c6dfba514c5c4
    - Deleted: 3da2bd307442e9edcda5b69cc4e343954465f814
  - DataFormats/ParticleFlowReco/src/classes.h:
    - Added: 42fd4a73bb03b04983328ba453595f9bcc3045b6
    - Modified: 71059c160a31476481edef20b1f3763d0a88045d, 3899652581347d9d54e2a3ca036c6dfba514c5c4
    - Deleted: 3da2bd307442e9edcda5b69cc4e343954465f814
  - DataFormats/ParticleFlowReco/src/classes_def_1.xml:
    - Deleted: 42fd4a73bb03b04983328ba453595f9bcc3045b6
    - Added: 3da2bd307442e9edcda5b69cc4e343954465f814
  - RecoParticleFlow/PFRecHitProducer/plugins/PFRecHitProducerTest.cc:
    - Added: e3a6018f0e8f209741533325c6311732ebf77a25
    - Modified: b09dd93946441de4b85491ddba4a25518565c3a7, 8cf3f0609deedcec4a59d0907553ec3d5dcaf2cf
    - Deleted: 302edcfe60dba1f5dc661e4d791d38d4a7515c17
  - DataFormats/ParticleFlowReco/src/alpaka/classes_serial_def.xml:
    - Added: 71059c160a31476481edef20b1f3763d0a88045d
    - Modified: 3899652581347d9d54e2a3ca036c6dfba514c5c4
    - Deleted: 3da2bd307442e9edcda5b69cc4e343954465f814

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42807/37226/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42807/37226/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

fllor commented:
Sorry for the delay everyone. I have implemented suggestions from @makortel and @fwyzard. I re-ran the validation and benchmarking to ensure that there are not unexpected consequences. There is a slight performance degradation for the GPU backend, but I since this module is extremely fast anyway, I think that does not matter. Please continue with the review.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42807/37232

- This PR adds an extra 120KB to repository

- Found files with invalid states:
  - DataFormats/ParticleFlowReco/src/classes_def_2.xml:
    - Deleted: 42fd4a73bb03b04983328ba453595f9bcc3045b6
    - Added: 3da2bd307442e9edcda5b69cc4e343954465f814
    - Modified: 2c956a278e8dd2f69e4b655c879f626fd22c767d
  - DataFormats/ParticleFlowReco/src/classes_1.h:
    - Deleted: 42fd4a73bb03b04983328ba453595f9bcc3045b6
    - Added: 3da2bd307442e9edcda5b69cc4e343954465f814
  - DataFormats/ParticleFlowReco/src/classes_2.h:
    - Deleted: 42fd4a73bb03b04983328ba453595f9bcc3045b6
    - Added: 3da2bd307442e9edcda5b69cc4e343954465f814
    - Modified: 2c956a278e8dd2f69e4b655c879f626fd22c767d
  - DataFormats/ParticleFlowReco/src/alpaka/classes_serial.h:
    - Added: 71059c160a31476481edef20b1f3763d0a88045d
    - Modified: 3899652581347d9d54e2a3ca036c6dfba514c5c4
    - Deleted: 3da2bd307442e9edcda5b69cc4e343954465f814
  - DataFormats/ParticleFlowReco/src/classes_def.xml:
    - Added: 42fd4a73bb03b04983328ba453595f9bcc3045b6
    - Modified: 71059c160a31476481edef20b1f3763d0a88045d, 3899652581347d9d54e2a3ca036c6dfba514c5c4
    - Deleted: 3da2bd307442e9edcda5b69cc4e343954465f814
  - DataFormats/ParticleFlowReco/src/classes.h:
    - Added: 42fd4a73bb03b04983328ba453595f9bcc3045b6
    - Modified: 71059c160a31476481edef20b1f3763d0a88045d, 3899652581347d9d54e2a3ca036c6dfba514c5c4
    - Deleted: 3da2bd307442e9edcda5b69cc4e343954465f814
  - DataFormats/ParticleFlowReco/src/classes_def_1.xml:
    - Deleted: 42fd4a73bb03b04983328ba453595f9bcc3045b6
    - Added: 3da2bd307442e9edcda5b69cc4e343954465f814
  - RecoParticleFlow/PFRecHitProducer/plugins/PFRecHitProducerTest.cc:
    - Added: e3a6018f0e8f209741533325c6311732ebf77a25
    - Modified: b09dd93946441de4b85491ddba4a25518565c3a7, 8cf3f0609deedcec4a59d0907553ec3d5dcaf2cf
    - Deleted: ed42d3a71aa73c082c62561ebbd3aa9a1175206d
  - DataFormats/ParticleFlowReco/src/alpaka/classes_serial_def.xml:
    - Added: 71059c160a31476481edef20b1f3763d0a88045d
    - Modified: 3899652581347d9d54e2a3ca036c6dfba514c5c4
    - Deleted: 3da2bd307442e9edcda5b69cc4e343954465f814

fllor commented on RecoParticleFlow/PFRecHitProducer/test/pfRecHitProducerAlpakaValidation.py:
```diff
@@ -0,0 +1,386 @@
+#####################################
+import sys
+import argparse
+parser = argparse.ArgumentParser(prog=f"{sys.argv[0]} {sys.argv[1]} --", description='Test and validation of PFRecHitProducer with Alpaka')
```
I do not have this PR available testing, but I changed it so that arguments to `cmsRun` are allowed. Arguments to the config still have to separated by `--`. I think this should also work when including the other PR.

makortel replied:
The https://github.com/cms-sw/cmssw/pull/42650 was merged last week and will thus be part of the 13_3_0_pre4 that should be built soon.

> Arguments to the config still have to separated by `--`.

Did you add this requirement in order to the configuration to work in some way with #42650 as well? I'm asking because with #42650 the `--` is not needed anymore to separate the `cmsRun` arguments from the configuration file arguments (basically the arguments before the configuration file are used for `cmsRun`, and the arguments after the configuration file are passed in `sys.argv`). I'd suggest to revise the argument handling of this configuration file e.g. after the pre4 has been built, but I'd be ok if that is done in a subsequent PR (since this configuration file is not used in any unit tests).

fllor replied:
Originally, I added this because otherwise arguments would always be interpreted as begin for `cmsRun`. Since `cmsRun` ignores anything after `--`, this was the easiest method I found to pass arguments to the config. With the new PR, the `--` is not needed anymore, I think it should still work if it is present. I agree that this should be changed at some point.

fllor replied:
I switched to `CMSSW_13_3_0_pre4`, where #42650 is available and implemented this suggestion.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42807/37237

- This PR adds an extra 124KB to repository

- Found files with invalid states:
  - DataFormats/ParticleFlowReco/src/classes_def_2.xml:
    - Deleted: 42fd4a73bb03b04983328ba453595f9bcc3045b6
    - Added: 3da2bd307442e9edcda5b69cc4e343954465f814
    - Modified: 2c956a278e8dd2f69e4b655c879f626fd22c767d
  - DataFormats/ParticleFlowReco/src/classes_1.h:
    - Deleted: 42fd4a73bb03b04983328ba453595f9bcc3045b6
    - Added: 3da2bd307442e9edcda5b69cc4e343954465f814
  - DataFormats/ParticleFlowReco/src/classes_2.h:
    - Deleted: 42fd4a73bb03b04983328ba453595f9bcc3045b6
    - Added: 3da2bd307442e9edcda5b69cc4e343954465f814
    - Modified: 2c956a278e8dd2f69e4b655c879f626fd22c767d
  - DataFormats/ParticleFlowReco/src/alpaka/classes_serial.h:
    - Added: 71059c160a31476481edef20b1f3763d0a88045d
    - Modified: 3899652581347d9d54e2a3ca036c6dfba514c5c4
    - Deleted: 3da2bd307442e9edcda5b69cc4e343954465f814
  - DataFormats/ParticleFlowReco/src/classes_def.xml:
    - Added: 42fd4a73bb03b04983328ba453595f9bcc3045b6
    - Modified: 71059c160a31476481edef20b1f3763d0a88045d, 3899652581347d9d54e2a3ca036c6dfba514c5c4
    - Deleted: 3da2bd307442e9edcda5b69cc4e343954465f814
  - DataFormats/ParticleFlowReco/src/classes.h:
    - Added: 42fd4a73bb03b04983328ba453595f9bcc3045b6
    - Modified: 71059c160a31476481edef20b1f3763d0a88045d, 3899652581347d9d54e2a3ca036c6dfba514c5c4
    - Deleted: 3da2bd307442e9edcda5b69cc4e343954465f814
  - DataFormats/ParticleFlowReco/src/classes_def_1.xml:
    - Deleted: 42fd4a73bb03b04983328ba453595f9bcc3045b6
    - Added: 3da2bd307442e9edcda5b69cc4e343954465f814
  - RecoParticleFlow/PFRecHitProducer/plugins/PFRecHitProducerTest.cc:
    - Added: e3a6018f0e8f209741533325c6311732ebf77a25
    - Modified: b09dd93946441de4b85491ddba4a25518565c3a7, 8cf3f0609deedcec4a59d0907553ec3d5dcaf2cf
    - Deleted: ed42d3a71aa73c082c62561ebbd3aa9a1175206d
  - DataFormats/ParticleFlowReco/src/alpaka/classes_serial_def.xml:
    - Added: 71059c160a31476481edef20b1f3763d0a88045d
    - Modified: 3899652581347d9d54e2a3ca036c6dfba514c5c4
    - Deleted: 3da2bd307442e9edcda5b69cc4e343954465f814

cmsbuild commented:
Pull request #42807 was updated. @fwyzard, @makortel, @jfernan2, @cmsbuild, @mandrenguyen can you please check and sign again.

fwyzard commented:
please test

makortel commented on RecoParticleFlow/PFRecHitProducer/test/BuildFile.xml:
```diff
@@ -0,0 +1,13 @@
+  <use name="FWCore/Framework"/>
+  <use name="FWCore/ParameterSet"/>
+  <use name="FWCore/Utilities"/>
+  <use name="alpaka"/>
```
Why is direct `alpaka` dependence needed?

fwyzard commented on RecoParticleFlow/PFRecHitProducer/interface/alpaka/CalorimeterDefinitions.h:
I think this file is only used by files under `RecoParticleFlow/PFRecHitProducer/plugins/alpaka/`.
Then, it could be made "private" and moved to `RecoParticleFlow/PFRecHitProducer/plugins/alpaka/` itself.

fwyzard replied:
Unless you plan to use it from some other package in the future ?

fwyzard replied:
The content of the file in principle looks good; the main concern is how much of the information is being duplicated from existing files, and whether it could be reused instead of duplicated.

fwyzard commented on RecoParticleFlow/PFRecHitProducer/interface/PFRecHitParamsRecord.h:
```diff
@@ -0,0 +1,14 @@
+
+class PFRecHitHCALParamsRecord : public edm::eventsetup::EventSetupRecordImplementation<PFRecHitHCALParamsRecord> {};
+
+class PFRecHitECALParamsRecord
+    : public edm::eventsetup::DependentRecordImplementation<PFRecHitECALParamsRecord,
+                                                            edm::mpl::Vector<EcalPFRecHitThresholdsRcd>> {};
```
Why not simply reuse the `EcalPFRecHitThresholdsRcd` instead of defining a new `PFRecHitECALParamsRecord` with the same IOVs ?
Do you foresee adding other dependencies in the future ?

fllor replied:
This record contains all the parameters that are required to apply cuts to ECAL hits to select which should become PFRecHits. Right now, only a single type of cut is implemented. I think it is possible that other types of cuts will be added in the future.

fwyzard commented on RecoParticleFlow/PFRecHitProducer/interface/PFRecHitParamsRecord.h:
```diff
@@ -0,0 +1,14 @@
+#include "FWCore/Framework/interface/DependentRecordImplementation.h"
+#include "CondFormats/DataRecord/interface/EcalPFRecHitThresholdsRcd.h"
+
+class PFRecHitHCALParamsRecord : public edm::eventsetup::EventSetupRecordImplementation<PFRecHitHCALParamsRecord> {};
```
If this record will always be used only to copy parameters from the python configuration to the event setup (to make them available on GPU), you could reuse the `JobConfigurationGPURecord` instead.

Currently it's defined in `HeterogeneousCore/CUDACore/interface/JobConfigurationGPURecord.h`, but we could move it to a place that is not dependent on CUDA.

fllor replied:
There is an effort to move these parameters from the configuration to the database. But it is not yet implemented in the serial version of the producer, so I also did not do it here.
Furthermore, similar to ECAL, other types of cut might be added in the future.

fwyzard replied:
ok

fwyzard commented on RecoParticleFlow/PFRecHitProducer/interface/PFRecHitTopologyRecord.h:
```diff
@@ -0,0 +1,17 @@
+                                       PFRecHitHCALTopologyRecord,
+                                       edm::mpl::Vector<HcalRecNumberingRecord, CaloGeometryRecord>> {};
+
+class PFRecHitECALTopologyRecord
+    : public edm::eventsetup::DependentRecordImplementation<PFRecHitECALTopologyRecord,
+                                                            edm::mpl::Vector<CaloGeometryRecord>> {};
```
You could reuse the `CaloGeometryRecord` instead of defining a new record.
Unless you plan to add other dependencies in the future.

fwyzard commented on RecoParticleFlow/PFRecHitProducer/plugins/alpaka/PFRecHitTopologyESProducer.cc:
Given the amount of HCAL-specific code, and the specialisations for `getNeighbourDetId()`, wouldn't it make more sense to just have two separate `ESProducer`s ?

fllor replied:
Personally, I prefer the current version, as it avoids code duplication. If you feel strongly about this, I will separate them.

fwyzard replied:
OK.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-cb4096/35245/summary.html
**COMMIT**: e34af3341b9fdb0386cd5346a4669c8988c7d7e3
**CMSSW**: CMSSW_13_3_X_2023-10-17-2300/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/42807/35245/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially removed 1 lines from the logs
 - Reco comparison results: 8 differences found in the comparisons
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 3357400
 - DQMHistoTests: Total failures: 6
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3357372
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 214 log files, 167 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_3_X_2023-10-17-2300+cb4096/59285/triggerResults): no differences found

fwyzard commented on RecoParticleFlow/PFRecHitProducer/plugins/alpaka/PFRecHitProducerKernel.dev.cc:
```diff
@@ -0,0 +1,198 @@
+        work_div_(cms::alpakatools::make_workdiv<Acc1D>(1, 1)) {}
+
+  template <typename CAL>
+  void PFRecHitProducerKernel<CAL>::prepareEvent(Queue& queue, const uint32_t num_recHits) {
```
I would suggest to merge `prepareEvent` with the constructor.
This would guarantee that the buffers are always properly initialised.

fwyzard commented on RecoParticleFlow/PFRecHitProducer/plugins/alpaka/PFRecHitProducerKernel.dev.cc:
```diff
@@ -0,0 +1,198 @@
+        // Use atomic operation to determine index of the PFRecHit to be constructed
+        // The index needs to be unique and consequtive across all threads in all blocks.
+        // This is achieved using the alpaka::hierarchy::Blocks argument.
+        const uint32_t j = alpaka::atomicAdd(acc, num_pfRecHits, 1u, alpaka::hierarchy::Blocks{});
```
Can you check if `alpaka::atomicInc` works correctly here ?
```suggestion
        const uint32_t j = alpaka::atomicInc(acc, num_pfRecHits, 0xffffffff, alpaka::hierarchy::Blocks{});
```

In principle it should generate more efficient code, at least for the CUDA backend.

For the CPU backend is shouldn't matter anyway, since operations with the `alpaka::hierarchy::Blocks{}` argument should decay to the standard, non-atomic version for the serial backend.

fllor replied:
It works, but there is no measurable performance difference on the CUDA backend. I will include this in the next commit.

fllor commented on RecoParticleFlow/PFRecHitProducer/plugins/alpaka/PFRecHitProducerAlpaka.cc:
```diff
@@ -0,0 +1,78 @@
+  using namespace ParticleFlowRecHitProducer;
+
+  template <typename CAL>
+  class PFRecHitProducerAlpaka : public global::EDProducer<> {
```
It would probably work, but I think it would be confusing, since we also have the serial implementation at: `RecoParticleFlow/PFClusterProducer/plugins/PFRecHitProducer.cc`. Then again, this file is in the `alpaka` subdirectory and the actual producers would be called `PFRecHitProducer{E,H}CAL` anyway. I do not really have a strong preference.

fwyzard replied:
An alternative could be `PFRecHitSoAProducer`, to highlight that the produced data formats are in SoA format.

fwyzard commented on RecoParticleFlow/PFRecHitProducer/plugins/alpaka/PFRecHitProducerKernel.dev.cc:
```diff
@@ -0,0 +1,198 @@
+    } else if (subdet == HcalEndcap) {
+      threshold = params.energyThresholds()[depth - 1 + HCAL::kMaxDepthHB];
+    } else {
+      printf("Rechit with detId %u has invalid subdetector %u!\n", detId, subdet);
```
would it make sense to add an explicit
```c++
      return false;
```
here ?

fwyzard commented on RecoParticleFlow/PFRecHitProducer/plugins/alpaka/PFRecHitProducerKernel.dev.cc:
```diff
@@ -0,0 +1,198 @@
+      return false;
+
+    // Reject ECAL recHits of bad quality
+    if ((ECAL::checkFlag(rh.flags(), ECAL::Flags::kOutOfTime) && rh.energy() > 2) ||
```
why `rh.energy() > 2` ?

fllor replied:
This is taken from https://github.com/fllor/cmssw/blob/pfRecHitProducerAlpaka/RecoParticleFlow/PFClusterProducer/interface/PFRecHitQTests.h#L458

fwyzard commented on RecoParticleFlow/PFRecHitProducer/plugins/alpaka/PFRecHitProducerKernel.dev.cc:
```diff
@@ -0,0 +1,198 @@
+        pfRecHits.y(i) = topology.positionY(denseId);
+        pfRecHits.z(i) = topology.positionZ(denseId);
+
+        for (uint32_t n = 0; n < 8; n++) {
+          pfRecHits.neighbours(i)(n) = -1;
+          const uint32_t denseId_neighbour = topology.neighbours(denseId)(n);
+          if (denseId_neighbour != 0xffffffff) {
+            const uint32_t pfRecHit_neighbour = denseId2pfRecHit[denseId_neighbour];
+            if (pfRecHit_neighbour != 0xffffffff)
+              pfRecHits.neighbours(i)(n) = (int32_t)pfRecHit_neighbour;
+          }
+        }
```
For the moment this is just an idea, not a suggestion to implement any changes.

How much time is spent in this kernel ?
If it's negligible, we don't need to change anything.
If it's significant, we could look into a couple of options:
  - inverting the loop, with an outer loop over `n` and an inner loop over `i`;
  - parallelising this loop using a 2D grid.

fllor replied:
This could probably be optimised. But the entire module is already extremely fast (~10us/event, maybe 20% of that in this loop), so I think this is probably not worth worrying about.

fwyzard commented on RecoParticleFlow/PFRecHitProducer/plugins/BuildFile.xml:
```diff
@@ -0,0 +1,33 @@
+  <use name="DataFormats/Common"/>
+  <use name="DataFormats/ParticleFlowReco"/>
+  <use name="CommonTools/ParticleFlow"/>
+  <use name="DQMServices/Core"/>
```
I think the dependency on DQM should not be necessary ?

fwyzard commented on RecoParticleFlow/PFRecHitProducer/plugins/BuildFile.xml:
```diff
@@ -0,0 +1,33 @@
+  <use name="FWCore/Framework"/>
+  <use name="FWCore/ParameterSet"/>
+  <use name="FWCore/Utilities"/>
+  <use name="alpaka"/>
```
The dependency on `alpaka` should not be needed.

hatakeyamak commented:
By the way, in the PR description, I think:
```
A producer that converts to new PFRecHitSoA format back to the existing PFRecHit collection format is included. It can be used for validation and to run down-stream modules, until they are adapted to the SoA themselves.
```
is more accurate with
```
A producer that converts to new PFRecHitSoA format back to the existing PFRecHit collection format is included. It will be used for validation and to run down-stream modules.
```
We will need this converter in order to keep the PFCluster to PFRecHits references, which will be used in various downstream modules.

fwyzard commented on RecoParticleFlow/PFRecHitProducer/interface/alpaka/CalorimeterDefinitions.h:
```diff
@@ -0,0 +1,203 @@
+      static constexpr bool positiveZ(uint32_t detId) { return detId & 0x4000; }
+
+      static constexpr uint32_t denseIndex(uint32_t detId) {
+        const unsigned short kxf[] = {
+            41, 51, 41, 51, 41, 51, 36, 51, 36, 51, 26, 51, 26, 51, 26, 51, 21, 51, 21, 51, 21, 51, 21, 51, 21,
+            51, 16, 51, 16, 51, 14, 51, 14, 51, 14, 51, 14, 51, 14, 51, 9,  51, 9,  51, 9,  51, 9,  51, 9,  51,
+            6,  51, 6,  51, 6,  51, 6,  51, 6,  51, 6,  51, 6,  51, 6,  51, 6,  51, 6,  51, 4,  51, 4,  51, 4,
+            51, 4,  51, 4,  56, 1,  58, 1,  59, 1,  60, 1,  61, 1,  61, 1,  62, 1,  62, 1,  62, 1,  62, 1,  62,
+            1,  62, 1,  62, 1,  62, 1,  62, 1,  62, 1,  61, 1,  61, 1,  60, 1,  59, 1,  58, 4,  56, 4,  51, 4,
+            51, 4,  51, 4,  51, 6,  51, 6,  51, 6,  51, 6,  51, 6,  51, 6,  51, 6,  51, 6,  51, 6,  51, 6,  51,
+            9,  51, 9,  51, 9,  51, 9,  51, 9,  51, 14, 51, 14, 51, 14, 51, 14, 51, 14, 51, 16, 51, 16, 51, 21,
+            51, 21, 51, 21, 51, 21, 51, 21, 51, 26, 51, 26, 51, 26, 51, 36, 51, 36, 51, 41, 51, 41, 51, 41, 51};
+
+        const unsigned short kdi[] = {
+            0,    10,   20,   30,   40,   50,   60,   75,   90,   105,  120,  145,  170,  195,  220,  245,  270,
+            300,  330,  360,  390,  420,  450,  480,  510,  540,  570,  605,  640,  675,  710,  747,  784,  821,
+            858,  895,  932,  969,  1006, 1043, 1080, 1122, 1164, 1206, 1248, 1290, 1332, 1374, 1416, 1458, 1500,
+            1545, 1590, 1635, 1680, 1725, 1770, 1815, 1860, 1905, 1950, 1995, 2040, 2085, 2130, 2175, 2220, 2265,
+            2310, 2355, 2400, 2447, 2494, 2541, 2588, 2635, 2682, 2729, 2776, 2818, 2860, 2903, 2946, 2988, 3030,
+            3071, 3112, 3152, 3192, 3232, 3272, 3311, 3350, 3389, 3428, 3467, 3506, 3545, 3584, 3623, 3662, 3701,
+            3740, 3779, 3818, 3857, 3896, 3935, 3974, 4013, 4052, 4092, 4132, 4172, 4212, 4253, 4294, 4336, 4378,
+            4421, 4464, 4506, 4548, 4595, 4642, 4689, 4736, 4783, 4830, 4877, 4924, 4969, 5014, 5059, 5104, 5149,
+            5194, 5239, 5284, 5329, 5374, 5419, 5464, 5509, 5554, 5599, 5644, 5689, 5734, 5779, 5824, 5866, 5908,
+            5950, 5992, 6034, 6076, 6118, 6160, 6202, 6244, 6281, 6318, 6355, 6392, 6429, 6466, 6503, 6540, 6577,
+            6614, 6649, 6684, 6719, 6754, 6784, 6814, 6844, 6874, 6904, 6934, 6964, 6994, 7024, 7054, 7079, 7104,
+            7129, 7154, 7179, 7204, 7219, 7234, 7249, 7264, 7274, 7284, 7294, 7304, 7314};
```
rather than duplicating these numbers, can you reuse them from `DataFormats/EcalDetId/interface/EEDetId.h` ?

fllor replied:
I do not know how, without modifying that file.

fwyzard replied:
We should really avoid the duplication.

However, as this goes beyond the gpu/alpaka code, maybe @cms-sw/reconstruction-l2 or @cms-sw/geometry-l2 may comment ?

fwyzard replied:
From a purely technical side, a simple change would be to mark those arrays as `HOST_DEVICE_CONSTANT`, like it was done in `Geometry/CommonTopologies/interface/SimplePixelTopology.h` .

fwyzard commented on RecoParticleFlow/PFRecHitProducer/interface/alpaka/CalorimeterDefinitions.h:
```diff
@@ -0,0 +1,203 @@
+    static constexpr int kSubdetectorBarrelId = HcalSubdetector::HcalBarrel;
+    static constexpr int kSubdetectorEndcapId = HcalSubdetector::HcalEndcap;
+
+    static constexpr uint32_t kMaxDepthHB = 4;
+    static constexpr uint32_t kMaxDepthHE = 7;
+    static constexpr uint32_t kFirstHBRing = 1;
+    static constexpr uint32_t kLastHBRing = 16;
+    static constexpr uint32_t kFirstHERing = 16;
+    static constexpr uint32_t kLastHERing = 29;
+    static constexpr uint32_t kMaxIPhi = 72;
```
Can these constants be taken from some existing files ?

fllor replied:
I cannot find a convenient place right now. The best one I see is https://github.com/fllor/cmssw/blob/ef3151a362db68a97ef10c4ff993af637cd163ef/Geometry/CaloTopology/src/HcalTopology.cc#L172-L175 but we cannot easily read from there.
@hatakeyamak, any ideas?

fwyzard replied:
Mhm, then I would suggest to move your definitions to a new file under `Geometry/CaloTopology/interface/`, and make both the alpaka PF code and the existing `HcalTopology.cc` code read them from there.

fllor replied:
I agree. These numbers are also duplicated elsewhere. But should this be part of this PR?

fwyzard replied:
For me it would be fine to do it in a separate PR.
In that case, please open an issue on GitHub and list there the changes that you propose to do in the separate PR.

hatakeyamak replied:
I don't have a good immediate suggestion. For maxDepth, it is probably better to read from:
https://github.com/cms-sw/cmssw/blob/master/Geometry/CaloTopology/src/HcalTopology.cc#L987
and actually this is in a sense not constant, because they changed after Phase 1 upgrade.
The current implementation is good for Run 3, but when we add it in official sequences in upcoming PRs, we will have to add some era dependence.
I also think it's fine to revisit in future PR (and have a github issue to keep a log.)

fwyzard commented:
> We will need this converter in order to keep the PFCluster to PFRecHits references, which will be used in various downstream modules.

I see... how much time do the SoA to legacy converters take ?

fwyzard commented on RecoParticleFlow/PFRecHitProducer/plugins/LegacyPFRecHitProducer.cc:
I see... how much time do the SoA to legacy converters take ?

fllor commented:
> > We will need this converter in order to keep the PFCluster to PFRecHits references, which will be used in various downstream modules.
> 
> I see... how much time do the SoA to legacy converters take ?

So far I never benchmarked this, because I also thought that it would be obsolete at some point.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42807/37251

- This PR adds an extra 36KB to repository

- Found files with invalid states:
  - DataFormats/ParticleFlowReco/src/classes_def_2.xml:
    - Deleted: 42fd4a73bb03b04983328ba453595f9bcc3045b6
    - Added: 3da2bd307442e9edcda5b69cc4e343954465f814
    - Modified: 2c956a278e8dd2f69e4b655c879f626fd22c767d
  - DataFormats/ParticleFlowReco/src/classes_1.h:
    - Deleted: 42fd4a73bb03b04983328ba453595f9bcc3045b6
    - Added: 3da2bd307442e9edcda5b69cc4e343954465f814
  - DataFormats/ParticleFlowReco/src/classes_2.h:
    - Deleted: 42fd4a73bb03b04983328ba453595f9bcc3045b6
    - Added: 3da2bd307442e9edcda5b69cc4e343954465f814
    - Modified: 2c956a278e8dd2f69e4b655c879f626fd22c767d
  - DataFormats/ParticleFlowReco/src/alpaka/classes_serial.h:
    - Added: 71059c160a31476481edef20b1f3763d0a88045d
    - Modified: 3899652581347d9d54e2a3ca036c6dfba514c5c4
    - Deleted: 3da2bd307442e9edcda5b69cc4e343954465f814
  - DataFormats/ParticleFlowReco/src/classes_def.xml:
    - Added: 42fd4a73bb03b04983328ba453595f9bcc3045b6
    - Modified: 71059c160a31476481edef20b1f3763d0a88045d, 3899652581347d9d54e2a3ca036c6dfba514c5c4
    - Deleted: 3da2bd307442e9edcda5b69cc4e343954465f814
  - RecoParticleFlow/PFRecHitProducer/interface/alpaka/CalorimeterDefinitions.h:
    - Added: c88f2562f6578cb248bebecfdb74b6a0ab47c7ea
    - Modified: ed262b10b72620186aa7866656cb16d28af72937, aa755b98ec04e729ed38dfdcc54cc030945c6548, c1bdf6f61b68b0aa840fa0defa4892fbf6d09486, d0edb2b7abee8772fefadf6e9cb491e0550b3b99, 40ed38b82330af3e012679723555174c143bc7ab, ed42d3a71aa73c082c62561ebbd3aa9a1175206d, ef8f121b69bc8e77994bb65187940d36cdb15855, 3443c5a1793d8a54df918a43440134131d54b637, 5946a3a7570d3835fe6bef2864600f115ee6d717, 8f43d090646ef0e80da8c9003acda91b5164bbf1
    - Deleted: 77f1a0d298857e6040e18dc0334fc66eafbf8c60
  - DataFormats/ParticleFlowReco/src/classes.h:
    - Added: 42fd4a73bb03b04983328ba453595f9bcc3045b6
    - Modified: 71059c160a31476481edef20b1f3763d0a88045d, 3899652581347d9d54e2a3ca036c6dfba514c5c4
    - Deleted: 3da2bd307442e9edcda5b69cc4e343954465f814
  - DataFormats/ParticleFlowReco/src/classes_def_1.xml:
    - Deleted: 42fd4a73bb03b04983328ba453595f9bcc3045b6
    - Added: 3da2bd307442e9edcda5b69cc4e343954465f814
  - RecoParticleFlow/PFRecHitProducer/plugins/PFRecHitProducerTest.cc:
    - Added: e3a6018f0e8f209741533325c6311732ebf77a25
    - Modified: b09dd93946441de4b85491ddba4a25518565c3a7, 8cf3f0609deedcec4a59d0907553ec3d5dcaf2cf
    - Deleted: ed42d3a71aa73c082c62561ebbd3aa9a1175206d
  - DataFormats/ParticleFlowReco/src/alpaka/classes_serial_def.xml:
    - Added: 71059c160a31476481edef20b1f3763d0a88045d
    - Modified: 3899652581347d9d54e2a3ca036c6dfba514c5c4
    - Deleted: 3da2bd307442e9edcda5b69cc4e343954465f814

cmsbuild commented:
Pull request #42807 was updated. @cmsbuild, @fwyzard, @jfernan2, @mandrenguyen, @makortel can you please check and sign again.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42807/37252

- This PR adds an extra 40KB to repository

- Found files with invalid states:
  - DataFormats/ParticleFlowReco/src/classes_def_2.xml:
    - Deleted: 42fd4a73bb03b04983328ba453595f9bcc3045b6
    - Added: 3da2bd307442e9edcda5b69cc4e343954465f814
    - Modified: 2c956a278e8dd2f69e4b655c879f626fd22c767d
  - DataFormats/ParticleFlowReco/src/classes_1.h:
    - Deleted: 42fd4a73bb03b04983328ba453595f9bcc3045b6
    - Added: 3da2bd307442e9edcda5b69cc4e343954465f814
  - DataFormats/ParticleFlowReco/src/classes_2.h:
    - Deleted: 42fd4a73bb03b04983328ba453595f9bcc3045b6
    - Added: 3da2bd307442e9edcda5b69cc4e343954465f814
    - Modified: 2c956a278e8dd2f69e4b655c879f626fd22c767d
  - DataFormats/ParticleFlowReco/src/alpaka/classes_serial.h:
    - Added: 71059c160a31476481edef20b1f3763d0a88045d
    - Modified: 3899652581347d9d54e2a3ca036c6dfba514c5c4
    - Deleted: 3da2bd307442e9edcda5b69cc4e343954465f814
  - DataFormats/ParticleFlowReco/src/classes_def.xml:
    - Added: 42fd4a73bb03b04983328ba453595f9bcc3045b6
    - Modified: 71059c160a31476481edef20b1f3763d0a88045d, 3899652581347d9d54e2a3ca036c6dfba514c5c4
    - Deleted: 3da2bd307442e9edcda5b69cc4e343954465f814
  - RecoParticleFlow/PFRecHitProducer/interface/alpaka/CalorimeterDefinitions.h:
    - Added: c88f2562f6578cb248bebecfdb74b6a0ab47c7ea
    - Modified: ed262b10b72620186aa7866656cb16d28af72937, aa755b98ec04e729ed38dfdcc54cc030945c6548, c1bdf6f61b68b0aa840fa0defa4892fbf6d09486, d0edb2b7abee8772fefadf6e9cb491e0550b3b99, 40ed38b82330af3e012679723555174c143bc7ab, ed42d3a71aa73c082c62561ebbd3aa9a1175206d, ef8f121b69bc8e77994bb65187940d36cdb15855, 3443c5a1793d8a54df918a43440134131d54b637, 5946a3a7570d3835fe6bef2864600f115ee6d717, 8f43d090646ef0e80da8c9003acda91b5164bbf1
    - Deleted: 77f1a0d298857e6040e18dc0334fc66eafbf8c60
  - DataFormats/ParticleFlowReco/src/classes.h:
    - Added: 42fd4a73bb03b04983328ba453595f9bcc3045b6
    - Modified: 71059c160a31476481edef20b1f3763d0a88045d, 3899652581347d9d54e2a3ca036c6dfba514c5c4
    - Deleted: 3da2bd307442e9edcda5b69cc4e343954465f814
  - DataFormats/ParticleFlowReco/src/classes_def_1.xml:
    - Deleted: 42fd4a73bb03b04983328ba453595f9bcc3045b6
    - Added: 3da2bd307442e9edcda5b69cc4e343954465f814
  - RecoParticleFlow/PFRecHitProducer/plugins/PFRecHitProducerTest.cc:
    - Added: e3a6018f0e8f209741533325c6311732ebf77a25
    - Modified: b09dd93946441de4b85491ddba4a25518565c3a7, 8cf3f0609deedcec4a59d0907553ec3d5dcaf2cf
    - Deleted: ed42d3a71aa73c082c62561ebbd3aa9a1175206d
  - RecoParticleFlow/PFRecHitProducer/plugins/alpaka/PFRecHitProducerAlpaka.cc:
    - Added: c1bdf6f61b68b0aa840fa0defa4892fbf6d09486
    - Modified: b09dd93946441de4b85491ddba4a25518565c3a7, 40ed38b82330af3e012679723555174c143bc7ab, ed42d3a71aa73c082c62561ebbd3aa9a1175206d, ef8f121b69bc8e77994bb65187940d36cdb15855, 77f1a0d298857e6040e18dc0334fc66eafbf8c60, e10fd2a939bc783eaf2fdccc1cd25ae05c18a219
    - Deleted: 420f5da5a546aa30a06c72cb955b1d1b57990775
  - DataFormats/ParticleFlowReco/src/alpaka/classes_serial_def.xml:
    - Added: 71059c160a31476481edef20b1f3763d0a88045d
    - Modified: 3899652581347d9d54e2a3ca036c6dfba514c5c4
    - Deleted: 3da2bd307442e9edcda5b69cc4e343954465f814

cmsbuild commented:
Pull request #42807 was updated. @jfernan2, @fwyzard, @makortel, @cmsbuild, @mandrenguyen can you please check and sign again.

fwyzard commented:
Looks good to me, thought maybe the history could be squashed to one or a few commits ?

makortel reviewed (commented):
Two minor comments that, as far as I'm concerned, can be addressed in a later PR.

fllor committed:
Add portable CaloRecHitSoA and PFRecHitSoA data format

The CaloRecHitSoA data format will serve as the base for ParticleFlow RecHit construction using Alpaka. It contains only those members necessary for ParticleFlow.
This is a temporary solution. As soon as official Alpaka-compatible SoAs become available from the HCAL/ECAL groups, this format becomes obsolete and will be replaced by those. This format also serves as a base for these two groups. The official formats should be super-sets of this one.
The PFRecHitSoA format will be used to store ParticleFlow RecHits in the Alpaka implementation. It contains exactly those members that are necessary for the ParticleFlow Cluster construction.

hatakeyamak commented:
So, it looks like @makortel and @fwyzard are happy with this PR.
Are other reviews happy with it?

fwyzard commented:
@fllor @hatakeyamak, I would prefer to squash the history to just one (or a few) commits, unless you prefer to keep it as it is.

hatakeyamak commented:
I am fine with as long as @fllor knows the simple way to do it.

fwyzard commented:
Although I've never used it, we have a brand new command line tool just for the job: `cms-squash-topic` .

fllor committed:
Add Alpaka producer that converts HCAL/ECAL hits to CaloRecHitSoA format

This commit adds a module that converts collections of CPU calorimeter hits (HBHERecHit or EcalRecHit) to the portable CaloRecHitSoA format. This step is necessary to process these hits in an Alpaka implementation of the ParticleFlow RecHit producer. Just like the CaloRecHitSoA format itself, this conversion is a temporary solution that will become obsolete, as soon as Alpaka-compatible SoAs are provided directly by the HCAL/ECAL groups.
The commit includes a configuration file, that executes the producer. This will serve as a basis for validation that will be introduced in future commits.

fllor committed:
Add ESProducers for Alpaka ParticleFlow RecHit producer

This commit adds two ESProducers that supply information necessary for PFRecHit construction:
1) Cut parameters: Calorimeter hits are subjected to certain quality cuts to decide if they should become ParticleFlow RecHits. One data structure is added to hold the parameters for these cuts. Two modules are provided to fill this data structure for HCAL and ECAL. Currently, only a a subset of the types of quality cuts that are supported by the CPU backend are supported in the new implementation. Further relevant types of cuts still need to be identified and implemented.
2) Topology information: This includes the positions (x,y,z) of each calorimeter cell and information about neighbouring cells. This information is necessary ParticleFlow RecHits from calorimeter hits and thus has to be supplied to the Alpaka ParticleFlow RecHit producer.

fllor committed:
Add Alpaka implementation of ParticleFlow RecHit producer

The PFRecHitSoAProducer constructs PFRecHits in the PFRecHitSoA format. Its input are calorimeter hits in the CaloRecHitSoA format and topology information and cut parameters, as supplied by the corresponding ESProducers.
The module is capable of consuming multiple CaloRecHitSoAs, as is required for ECAL, where barrel and endcap hits are supplied separately.

fllor committed:
Add conversion from PFRecHitSoA to legacy PFRecHit collection

This commit adds a producer that reads the PFRecHitSoA from the Alpaka module and produces a PFRecHit collection in the same format as the CPU PFRecHitProducer. This conversion is necessary as long as not all consumers of the PFRecHits are updated to use the portable SoA format. Since some downstream modules use PFCluster to PFRecHit references, this conversion is going to remain necessary for the foreseeable future. It can also be used to validate the Alpaka implementation of the PFRecHitProducer.

fllor committed:
Add validation for Alpaka PFRecHitProducer

Add DQMEDAnalyzer that compares two sets of PFRecHits. It can read both the legacy PFRecHit collection or the Alpaka PFRecHitSoA format. Since the construction of PFRecHits does not involve any numerical computation, PFRecHits are not subject to numerical fluctuations. Therefore the validation simply tests all members for equality. The analyzer does not produce plots, but rather prints its findings to the log.
When calorimeter hits become available with Alpaka, this strict validation might have to be relaxed, as calorimeter reconstruction is subject to numerical fluctuations.

The validation can be performed using the pfRecHitProducerAlpakaValidation.py configuration. Possible command-line options are:
  --cal CAL             Calorimeter type. Possible options: HCAL, ECAL. Default: HCAL
  --backend BACKEND     Alpaka backend. Possible options: CPU, GPU, auto. Default: auto
  --threads THREADS     Number of threads. Default: 8
  --debug               Dump PFRecHits from both inputs for first event (implies `--threads 1`)
  --synchronise         Put synchronisation point at the end of Alpaka modules (for benchmarking asynchronous execution)

Expected output:
$ cmsRun RecoParticleFlow/PFRecHitProducer/test/pfRecHitProducerAlpakaValidation.py --cal hcal --backend gpu --threads 1
...
PFRecHitProducerTest[Legacy vs Alpaka] has compared 100 events and found 0 problems: [0, 0, 0, 0, 0]
PFRecHitProducerTest[Alpaka vs Legacy-from-Alpaka] has compared 100 events and found 0 problems: [0, 0, 0, 0, 0]
PFRecHitProducerTest[Legacy vs Legacy-from-Alpaka] has compared 100 events and found 0 problems: [0, 0, 0, 0, 0]

The module performs three validations:
1) Legacy vs Alpaka: This compares the PFRecHit collection from the legacy CPU backend to the PFRecHitSoA format from the Alpaka module. This is the main validation of the Alpaka implementation.
2) Alpaka vs Legacy-from-Alpaka: This compares the PFRecHitSoA format from the Alpaka module to the legacy PFRecHit collection that was constructed from the PFRecHitSoA. This validates the conversion module.
3) Legacy vs Legacy-from-Alpaka: This compares the PFRecHit collection from the legacy CPU backend to the PFRecHit collection that was constructed from the PFRecHitSoA. This is mainly a consistency check and it helps to valdiate the validation module.
In case of problems, exactly two of these validations should fail.

Similar output is expected for every combination of calorimeter type and Alpaka backend. When using multiple threads, there will be multiple such outputs at the end. In case the validation fails, the numbers in brackets indicate which type of error occured, which is useful for debugging.

Right now, this module tests PFRecHits for exact equality, including floating point numbers. This works, since PFRecHit construction does not introduce numerical fluctuations. Later on, when also transitioning the calorimeter reconstruction to Alpaka, fluctuations are expected in the calorimeter hits and this simple validation of the PFRecHits will not suffice.
Therefore, the equality tests of the "energy" and "time" members are optional. Additionally, 2D histograms are created that can be used to visually validate the results. The histograms can be drawn by running:

$ cmsRun RecoParticleFlow/PFRecHitProducer/test/pfRecHitProducerAlpakaValidation.py
$ cmsRun RecoParticleFlow/PFRecHitProducer/test/harvestValidationPlot.py
$ root -q RecoParticleFlow/PFRecHitProducer/test/drawValidationPlot.C

This will produce two plots "pfRecHit_energy.png" and "pfRecHit_time.png". At the moment, they will show perfect diagonals.

fllor commented on DataFormats/ParticleFlowReco/interface/CaloRecHitHostCollection.h:
```diff
@@ -0,0 +1,13 @@
+  using CaloRecHitHostCollection = PortableHostCollection<CaloRecHitSoA>;
+
+}  // namespace reco
+
```
It seems this is not yet available in `CMSSW_13_3_0_pre4`?

fwyzard replied:
Correct, that was merged a few days later; we can leave it for a follow up PR.

cmsbuild commented:
Pull request #42807 was updated. @mandrenguyen, @fwyzard, @jfernan2, @makortel, @cmsbuild can you please check and sign again.

fwyzard commented:
+heterogeneous

fwyzard commented:
please test

hatakeyamak commented:
Thank you. So, in terms of signatures, we still need it from reconstruction and pf?

fwyzard commented:
> Thank you. So, in terms of signatures, we still need it from reconstruction and pf?

Technically, only from @cms-sw/reconstruction-l2, while @cms-sw/pf-l2 is only "for information" and does not require a signature.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-cb4096/35413/summary.html
**COMMIT**: 3f4d1235b92a1151179071f3e1396ceb34d50304
**CMSSW**: CMSSW_13_3_X_2023-10-24-2300/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/42807/35413/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially added 69 lines to the logs
 - Reco comparison results: 29 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3351458
 - DQMHistoTests: Total failures: 1073
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3350363
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 213 log files, 166 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_3_X_2023-10-24-2300+cb4096/59415/triggerResults): no differences found

mandrenguyen commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @antoniovilela, @rappoccio, @sextonkennedy (and backports should be raised in the release meeting by the corresponding L2)

antoniovilela commented:
+1
